runtime/lua/vim/shared.lua

```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
   return s:gsub('[%(%)%.%%%+%-%*%?%[%]%^%$]', '%%%1')
 end
 
+--- Type checking validation function
+---
+--- Examples:
+--- <pre>
+---  validate({ arg={ { 'foo' }, 'table' }})                          --> Nop
+---  validate({ arg={ 1, 'table' } })                                 --> error("'arg' is expected table, got number")
+---  validate({ arg1={ { 'foo' }, 'table' }, arg2={ 1, 'string' } })  --> error("'arg2' is expected string, got number")
+--- </pre>
+---
+---@param ... Table or list of table. That table is "argument_name = { validation_target, type_name (, whether nil is allowed) }".
+--- The following can be used as type_names:
+---     - table or tbl
+---     - string or str
+---     - number or num
+---     - boolean or bool
+---     - function or func
+---     - nil
+---     - thread
+---     - userdata
+function vim.validate(...)
```

Comment:
I don't see a reason for this to be variadic (`...`). Seems like a table can/should be required.
```suggestion
function vim.validate(opt)
  assert(type(opt) == 'table')
```